src/main/java/godaddy/registry/jtoolkit2/session/SessionProperties.java

```diff
@@ -169,4 +169,18 @@
      * @return a boolean flag where true means XML output will have namespace prefix
      */
     boolean needOutputNamespacePrefixInXml();
+
+    /**
+     * The bytes representation of the KeyStore file from which private key and public
+     * certificate data will be read in order to establish a connection with
+     * the EPP server.
+     */
+    byte[] getKeyStoreFileBytes();
+
+    /**
+     * The bytes representation of the KeyStore file from which Certification Authority
+     * data will be read in order to verify the authenticity of the key
+     * in the accompanying user keystore.
+     */
+    byte[] getTrustStoreFileBytes();
```

Comment:
It does not appear that we'll need these two new methods, because

1. It's not a good idea to expose getter to return the content of keystore/trust-store (even though they're protected by password).
2. With the existing methods - `getKeyStoreFilename()` and `getTrustStoreFilename()`, it should be sufficient to allow your version of classes to read a URI of location, and then load the bytes internally.

    The name of these two existing methods may not be accurate, but from their javadoc, it does indicate they are used to retrieve "the location of" keystore/trust-store.

    Using Spring framework's `ResourceLoader` as [an example](https://dzone.com/articles/working-with-resources-in-spring), we could treat the returned `String` as a logical URL - it could be some resource in classpath, file system, or could be loaded via HTTP(S)/SFTP and etc.

Therefore, it would be better to remove these two new methods from this interface.

Reply:
It's fine if you would like to store the keystore/strustore in database. However, the `SessionProperties` interface does not need to add these two new methods just because of that.

Below is some pseudo code to illustrate a few options:

* Option 1 - loading the bytes when `configure(SessionProperties properties)` is called on the subclass, based on the (primary) identifiers of session-properties

    Given that it's one-to-one mapping between the `SessionProperties` and keystore/truststore, I guess it's fair to assume that there is some identifier of each `SessionProperties` (e.g., primary key of this row in your database). Let's say, `properties.getClientID()` is the PK. It's possible to put some code in the `configure(...)` method:

    ```java
	public class MyTLSSession extends TLSSession {

		// ......

		@Override
		public void configure(SessionProperties properties) throws SessionConfigurationException {

			// ...
			this.username = properties.getClientID();
			bytes[] keystoreBytes = loadKeyStoreFromDatabase(this.username);
			bytes[] truststoreBytes = loadTrustStoreFromDatabase(this.username);

			// continue to construct your version of TLSContext by passing the bytes ...
		}

	}
    ```

* Option 2 - use runtime type check and call the methods on the instance of subclass

    In case, for some reason, you have loaded the bytes on your own version of `SessionProperties`, and just need a way to expose it to subclass of `TLSSession`, it can still be done like

    ```java
	public class MyTLSSession extends TLSSession {

		// ......

		@Override
		public void configure(SessionProperties properties) throws SessionConfigurationException {

			// ...
			bytes[] keystoreBytes;
			bytes[] truststoreBytes
			if (properties instanceof MySessionProperties) {
				MySessionProperties mp = (MySessionProperties) properties;
				keystoreBytes = mp.getKeyStoreFileBytes();
				truststoreBytes = mp.getTrustStoreFileBytes();
			} else {
				throw new IllegalStateException(); // should never happen...
			}

			// continue to construct your version of TLSContext by passing the bytes ...
		}

	}
    ```

Reply:
Thanks for detailed information. Yes, I have removed new methods from `SessionProperties` and used locally in my own version of `SessionProperties`.